File: GameLibrary/src/cc/lib/game/AMultiPhaseAnimation.java
package cc.lib.game;

import java.util.List;

/**
 * Created by Chris Caron on 8/18/21.
 */
public abstract class AMultiPhaseAnimation<T> extends AAnimation<T> {

    private long [] durations;

    protected AMultiPhaseAnimation(long [] durations) {
        super(Utils.sum(durations));
        this.durations = durations;
    }

    protected AMultiPhaseAnimation(long duration) {
        super(duration);
        this.durations = Utils.toLongArray(duration);
    }

    public AMultiPhaseAnimation(long durationMSecs, int repeats) {
        super(durationMSecs, repeats);
        this.durations = Utils.toLongArray(durationMSecs);
    }

    public AMultiPhaseAnimation(long durationMSecs, int repeats, boolean oscillateOnRepeat) {
        super(durationMSecs, repeats, oscillateOnRepeat);
        this.durations = Utils.toLongArray(durationMSecs);
    }

    /**
     * Draw a phase of a complete animation. Phase will be a value between [0-durations.length)
     *
     * @param g
     * @param position value between 0-1
     * @param phase
     */
    protected abstract void drawPhase(T g, float position, int phase);

    @Override
    protected void draw(T g, float position, float dt) {
        // phase 1: make skull fade in over the actor
        long dur=0;
        for (int i=0; i<durations.length; i++) {
            long d = durations[i]+dur;
            if (getElapsedTime() < d) {
                float pos = (getElapsedTime()-dur) / (float)durations[i];
                drawPhase(g, pos, i);
                break;
            }
            dur+=durations[i];
        }
    }

    public void setDuration(int phase, long duration) {
        durations[phase] = duration;
        super.setDuration(Utils.sum(durations));
    }

    public void setDurations(long ... durations) {
        this.durations = durations;
        setDuration(Utils.sum(durations));
    }

    public void setDurations(List<Long> durations) {
        long [] ldurs = new long[durations.size()];
        for (int i=0; i<durations.size(); i++)
            ldurs[i] = durations.get(i);
        setDurations(ldurs);
    }
}

File: GameLibrary/src/cc/lib/game/AMultiPhaseAnimation.kt
package cc.lib.game

/**
 * Created by Chris Caron on 8/18/21.
 */
abstract class AMultiPhaseAnimation<T> : AAnimation<T> {
	private var durations: LongArray
	private var lastPhase = -1

	protected constructor(durations: LongArray) : super(Utils.sum(durations)) {
		this.durations = durations
	}

	protected constructor(duration: Long) : super(duration) {
		durations = Utils.toLongArray(duration)
	}

	constructor(durationMSecs: Long, repeats: Int) : super(durationMSecs, repeats) {
		durations = Utils.toLongArray(durationMSecs)
	}

	constructor(durationMSecs: Long, repeats: Int, oscillateOnRepeat: Boolean) : super(durationMSecs, repeats, oscillateOnRepeat) {
		durations = Utils.toLongArray(durationMSecs)
	}

	/**
	 * Draw a phase of a complete animation. Phase will be a value between [0-durations.length)
	 *
	 * @param g
	 * @param positionInPhase value between 0-1
	 * @param positionInAnimation value between 0-1
	 * @param phase
	 */
	protected abstract fun drawPhase(g: T, positionInPhase: Float, positionInAnimation: Float, phase: Int)
	override fun draw(g: T, position: Float, dt: Float) {
		var dur: Long = 0
		for (i in durations.indices) {
			val d = durations[i] + dur
			val elapsedTime = elapsedTime
			if (elapsedTime < d) {
				val pos = (elapsedTime - dur) / durations[i].toFloat()
				if (i != lastPhase) {
					lastPhase = i
					onPhaseStarted(g, i)
				}
				drawPhase(g, pos, position, i)
				break
			}
			dur += durations[i]
		}
	}

	fun setDuration(phase: Int, duration: Long) {
		durations[phase] = duration
		this.duration = durations.sum()
	}

	fun setDurations(durations: List<Long>) {
		this.durations = durations.toLongArray()
		duration = durations.sum()
	}

	protected open fun onPhaseStarted(g: T, phase: Int) {}
}